docs/learn/multi-format-reader.md

```diff
@@ -0,0 +1,263 @@
+
+### Data post processing
+```python
+   self.post_process()
+```
+In case there is the need for any post-processing on the data and/or config dictionary _after_ they have been read, the `post_process` method can be implemented. For example, this can be helpful if there are multiple entities of a given NX_CLASS (for example, multiple detectors) on the same level and the config dict shall be set up to fill the template with all of these entities.
+
+### Filling the template from the read-in data
+```python
+    if self.config_dict:
+        suppress_warning = kwargs.pop("suppress_warning", False)
+        template.update(
+            fill_from_config(
+                self.config_dict,
+                self.get_entry_names(),
+                self.callbacks,
+                suppress_warning=suppress_warning,
+            )
+        )
+
+    return template
+```
+As a last step, the template is being filled from the config dict using the data. If there is more than one entry, the `get_entry_names` method must be implemented, which shall return a list of all entry names. The `fill_from_config` method iterates through all of the them and replaces the generic `/ENTRY/` in the config file by keys of the form `/ENTRY[my-entry]/` to fill the template.
+
+Here, we are using **callbacks**, which are used to bring in data based on `@`-prefixes in the config file. These are defined in the reader's ``__init__`` call using the `pynxtools.dataconverter.readers.multi.ParseJsonCallbacks` class:
+```python
+self.callbacks = ParseJsonCallbacks(
+    attrs_callback=self.get_attr,
+    data_callback=self.get_data,
+    eln_callback=self.get_eln_data,
+    dims=self.get_data_dims,
+)
+```
+The `ParseJsonCallbacks` class has an attribute called `special_key_map` that makes use of these callbacks to populate the template based on the starting prefix of the config dict value:
+```python
+self.special_key_map = {
+    "@attrs": attrs_callback if attrs_callback is not None else self.identity,
+    "@link": link_callback if link_callback is not None else self.link_callback,
+    "@data": data_callback if data_callback is not None else self.identity,
+    "@eln": eln_callback if eln_callback is not None else self.identity,
+}
+```
+That means, if the config file has an entry ```{"/ENTRY/title": "@attrs:metadata/title"}```, the `get_attr` method of the reader gets called and should return an attribute from the given path, i.e., in this case from `metadata/title`.
+
+By default, the MultiFormatReader supports the following special prefixes:
+
+- `@attrs`: To get metadata from the read-in experiment file(s). You need to implement the `get_attr` method in the reader.
+- `@data`: To get measurement data from the read-in experiment file(s). You need to implement the `get_data` method in the reader.

```

Comment:
The destinction between data and metadata is somewhat arbitrary here. The reason to have both of these is for files/object types that require different methods to access different parts of the object, e.g. the xarrays. Maybe woth explaining this here.

Reply:
good point, added